fabric-input-events-v1/src/main/java/net/fabricmc/fabric/impl/client/input/InputCallbacksImpl.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.client.input;
+
+import java.util.Map;
+
+import org.lwjgl.glfw.GLFW;
+import org.lwjgl.glfw.GLFWDropCallback;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+
+import net.minecraft.client.options.KeyBinding;
+import net.minecraft.client.util.InputUtil;
+import net.minecraft.client.util.InputUtil.Key;
+
+import net.fabricmc.fabric.api.client.input.v1.FabricMouse;
+import net.fabricmc.fabric.api.event.client.input.ClientInputEvents;
+import net.fabricmc.fabric.api.event.client.input.KeyEvent;
+import net.fabricmc.fabric.api.event.client.input.KeybindEvent;
+import net.fabricmc.fabric.api.event.client.input.MouseButtonEvent;
+import net.fabricmc.fabric.api.event.client.input.MouseMoveEvent;
+import net.fabricmc.fabric.api.event.client.input.MouseScrollEvent;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.InputUtilTypeAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.KeyEventAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.KeybindEventAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.KeyBindingAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.MouseButtonEventAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.MouseMoveEventAccessor;
+import net.fabricmc.fabric.mixin.event.input.client.accessor.GenericMouseEventAccessor;
+
+public final class InputCallbacksImpl {
+	private static final KeyEvent sharedKeyEvent = new KeyEvent(-1, -1, GLFW.GLFW_RELEASE, 0);
+	private static final KeybindEvent sharedKeybindEvent = new KeybindEvent(-1, -1, GLFW.GLFW_RELEASE, 0, null, null);
+	private static final MouseMoveEvent sharedMouseMoveEvent = new MouseMoveEvent(0.0, 0.0, 0.0, 0.0);
+	private static final MouseButtonEvent sharedMouseButtonEvent = new MouseButtonEvent(0, GLFW.GLFW_RELEASE, 0, null);
+	private static final MouseScrollEvent sharedMouseScrollEvent = new MouseScrollEvent(0.0, 0.0);
+
+	private static KeyEvent createKeyEvent(int code, int scancode, int action, int modKeys) {
+		KeyEventAccessor accessor = (KeyEventAccessor) (Object) sharedKeyEvent;
```

Comment:
This can't be worse. You create something people believe are pojo; they are not. You just fuck up people's assumptions this way, especially that you don't document any of these :angry:

Reply:
I've already talked about this, if we aren't going to allow anything other than multi-arg lambda's, then we'll just have to use those.